File: src/cards/parse_bang_cards.cpp
#include <regex>
#include <sstream>
#include <fstream>
#include <iostream>

#include <yaml-cpp/yaml.h>
#include <fmt/core.h>

void set_expansion(YAML::Node &card, const std::string &str) {
    if (auto expansion = card["expansion"]) {
        expansion = fmt::format("{} {}", str, expansion.as<std::string>());
    } else {
        expansion = str;
    }
}

bool is_hidden(const YAML::Node &card) {
    if (auto &tags = card["tags"]) {
        for (const auto &str : tags) {
            if (str.as<std::string>() == "hidden") {
                return true;
            }
        }
    }
    return false;
}

std::string parse_sign(const std::string &sign) {
    static const std::regex sign_regex {
        "^\\s*([\\w\\d]+)\\s*(\\w+)\\s*$"
    };
    std::smatch match;
    if (std::regex_match(sign, match, sign_regex)) {
        return fmt::format("card_suit::{0}, card_rank::rank_{1}", match.str(2), match.str(1));
    } else {
        throw std::runtime_error(fmt::format("Invalid sign string: {}", sign));
    }
}

std::string enum_flags(const std::string &sign, const std::string &enum_name) {
    std::string ret;
    size_t begin = sign.find_first_not_of(" \t");
    while (begin != std::string::npos) {
        size_t end = sign.find_first_of(" \t", begin);
        if (end == std::string::npos) {
            end = sign.size();
        }
        if (!ret.empty()) {
            ret += " | ";
        }
        ret += fmt::format("{}::{}", enum_name, sign.substr(begin, end - begin));
        begin = sign.find_first_not_of(" \t", end);
    }
    return ret;
}

void parse_effects(std::ostream &out, const YAML::Node &list, const std::string &name) {
    static const std::regex effect_regex {
        "^\\s*(\\w+)" // type
        "(?:\\s*\\((-?\\d+)\\))?" // effect_value
        "(?:\\s*(\\w+)\\s*)?" // target
        "([\\w\\s]*?)" // player_filter
        "(?:\\s*\\|\\s*([\\w\\s]+))?\\s*$", // card_filter
    };
    out << fmt::format(",\n      .{} {{\n", name);
    for (auto it = list.begin(); it != list.end();) {
        const std::string effect = it->as<std::string>();
        std::smatch match;
        if (!std::regex_match(effect, match, effect_regex)) {
            throw std::runtime_error(fmt::format("Invalid effect string: {}", effect));
        }
        auto type = match.str(1);
        auto effect_value = match.str(2);
        auto target = match.str(3);
        auto player_filter = match.str(4);
        auto card_filter = match.str(5);

        out << "        {\n";

        if (!target.empty()) {
            out << fmt::format("          .target {{target_type::{}}},\n", target);
        }
        if (!player_filter.empty()) {
            if (target != "player" && target != "conditional_player" && target != "card") {
                throw std::runtime_error(fmt::format("Invalid effect string: {0}\nPlayer filter not allowed with {1}", effect, target));
            }
            out << fmt::format("          .player_filter {{{}}},\n", enum_flags(player_filter, "target_player_filter"));
        }
        if (!card_filter.empty()) {
            if (target != "card") {
                throw std::runtime_error(fmt::format("Invalid effect string: {0}\nCard filter not alowed with {1}", effect, target));
            }
            out << fmt::format("          .card_filter {{{}}},\n", enum_flags(card_filter, "target_card_filter"));
        }
        if (!effect_value.empty()) {
            out << fmt::format("          .effect_value {{{}}},\n", effect_value);
        }
        out << fmt::format("          .type {{effect_type::{}}}\n        }}", type);
        if (++it == list.end()) {
            out << "\n";
        } else {
            out << ",\n";
        }
    }
    out << "      }";
}

void parse_effect_simple(std::ostream &out, const YAML::Node &list,
    const std::string &name, const std::string &value_name, const std::string &type_name)
{
    static const std::regex effect_regex {
        "^\\s*(\\w+)" // type
        "(?:\\s*\\((-?\\d+)\\))?\\s*$" // effect_value
    };

    out << fmt::format(",\n      .{} {{\n", name);
    for (auto it = list.begin(); it != list.end();) {
        const std::string effect = it->as<std::string>();
        std::smatch match;
        if (!std::regex_match(effect, match, effect_regex)) {
            throw std::runtime_error(fmt::format("Invalid {0} string: {1}", name, effect));
        }

        std::string type = match.str(1);
        std::string value = match.str(2);

        out << "        {\n";
        if (!value.empty()) {
            out << fmt::format("          .{0} {{{1}}},\n", value_name, value);
        }
        out << fmt::format("          .type {{{0}::{1}}}\n        }}", type_name, type);
        if (++it == list.end()) {
            out << "\n";
        } else {
            out << ",\n";
        }
    }
    out << "      }";
}

void parse_all_effects(std::ostream &out, YAML::Node card) {
    try {
        out << fmt::format("    {{\n      .name {{\"{}\"}}", card["name"].as<std::string>());
        if (auto image = card["image"]) {
            out << fmt::format(",\n      .image {{\"{}\"}}", image.as<std::string>());
        }
        if (auto effects = card["effects"]) {
            parse_effects(out, effects, "effects");
        }
        if (auto responses = card["responses"]) {
            parse_effects(out, responses, "responses");
        }
        if (auto optional = card["optional"]) {
            parse_effects(out, optional, "optionals");
        }
        if (auto equip = card["equip"]) {
            parse_effect_simple(out, equip, "equips", "effect_value", "equip_type");
        }
        if (auto tags = card["tags"]) {
            parse_effect_simple(out, tags, "tags", "tag_value", "tag_type");
        }
        if (auto expansion = card["expansion"]) {
            out << fmt::format(",\n      .expansion {{{}}}",
                enum_flags(expansion.as<std::string>(), "card_expansion_type"));
        }
        if (auto deck = card["deck"]) {
            out << fmt::format(",\n      .deck {{card_deck_type::{}}}", deck.as<std::string>());
        }
        if (auto modifier = card["modifier"]) {
            out << fmt::format(",\n      .modifier {{card_modifier_type::{}}}", modifier.as<std::string>());
        }
        if (auto mth_effect = card["mth_effect"]) {
            out << fmt::format(",\n      .mth_effect {{mth_type::{}}}", mth_effect.as<std::string>());
        }
        if (auto mth_response = card["mth_response"]) {
            out << fmt::format(",\n      .mth_response {{mth_type::{}}}", mth_response.as<std::string>());
        }
        if (auto equip_target = card["equip_target"]) {
            out << fmt::format(",\n      .equip_target {{target_player_filter::{}}}", equip_target.as<std::string>());
        }
        if (auto color = card["color"]) {
            out << fmt::format(",\n      .color {{card_color_type::{}}}", color.as<std::string>());
        }
    } catch (const std::runtime_error &error) {
        throw std::runtime_error(fmt::format("Error in card {0}:\n{1}", card["name"].as<std::string>(), error.what()));
    }
}

void parse_file(std::ostream &out, const YAML::Node &data) {
    std::vector<YAML::Node> hidden_cards;

    out <<
        "// AUTO GENERATED FILE\n\n"
        "#include \"game/card_data.h\"\n\n"
        "namespace banggame {\n\n"
        "using namespace enums::flag_operators;\n\n"
        "const all_cards_t all_cards {\n";
    
    out << "  .deck {\n";
    for (YAML::Node card : data["main_deck"]) {
        card["deck"] = "main_deck";
        for (const YAML::Node &sign : card["signs"]) {
            parse_all_effects(out, card);
            out << fmt::format(",\n      .sign {{{}}}\n    }},\n", parse_sign(sign.as<std::string>()));
        }
    }
    out << "  },\n";

    out << "  .characters {\n";
    for (YAML::Node card : data["character"]) {
        card["deck"] = "character";
        parse_all_effects(out, card);
        out << "\n    },\n";
    }
    out << "  },\n";

    out << "  .goldrush {\n";
    for (YAML::Node card : data["goldrush"]) {
        set_expansion(card, "goldrush");
        card["deck"] = "goldrush";
        int count = card["count"].as<int>(1);
        for (int i=0; i<count; ++i) {
            if (is_hidden(card)) {
                hidden_cards.push_back(card);
            } else {
                parse_all_effects(out, card);
                out << "\n    },\n";
            }
        }
    }
    out << "  },\n";

    out << "  .highnoon {\n";
    for (YAML::Node card : data["highnoon"]) {
        set_expansion(card, "highnoon");
        card["deck"] = "highnoon";
        if (is_hidden(card)) {
            hidden_cards.push_back(card);
        } else {
            parse_all_effects(out, card);
            out << "\n    },\n";
        }
    }
    out << "  },\n";

    out << "  .fistfulofcards {\n";
    for (YAML::Node card : data["fistfulofcards"]) {
        set_expansion(card, "fistfulofcards");
        card["deck"] = "fistfulofcards";
        if (is_hidden(card)) {
            hidden_cards.push_back(card);
        } else {
            parse_all_effects(out, card);
            out << "\n    },\n";
        }
    }
    out << "  },\n";

    out << "  .wildwestshow {\n";
    for (YAML::Node card : data["wildwestshow"]) {
        set_expansion(card, "wildwestshow");
        card["deck"] = "wildwestshow";
        if (is_hidden(card)) {
            hidden_cards.push_back(card);
        } else {
            parse_all_effects(out, card);
            out << "\n    },\n";
        }
    }
    out << "  },\n";

    out << "  .specials {\n";
    for (YAML::Node card : data["specials"]) {
        if (is_hidden(card)) {
            hidden_cards.push_back(card);
        } else {
            parse_all_effects(out, card);
            out << "\n    },\n";
        }
    }
    out << "  },\n";

    out << "  .hidden {\n";
    for (YAML::Node card : hidden_cards) {
        parse_all_effects(out, card);
        out << "\n    },\n";
    }
    out << "  },\n";

    out << "};\n\n}\n";
}

int main(int argc, char **argv) {
    if (argc != 3) {
        std::cout << fmt::format("Usage: {} bang_cards.yml bang_cards.cpp\n", argv[1]);
        return 1;
    }

    std::stringstream out;
    parse_file(out, YAML::LoadFile(argv[1]));

    std::ofstream ofs(argv[2]);
    ofs << out.str();

    return 0;
}

File: src/cards/parse_bang_cards.py
#!/usr/bin/env python3

import re
import sys
import yaml
from cpp_generator import CppEnum, print_cpp_file

def is_hidden(card):
    return 'tags' in card and any(value == 'hidden' for value in card['tags'])

def parse_sign(sign):
    match = re.match(r'^\s*([\w\d]+)\s*(\w+)\s*$', sign)
    if match:
        return {
            'suit': CppEnum('card_suit', match.group(2)),
            'rank' : CppEnum('card_rank', 'rank_' + match.group(1))
        }
    else:
        raise RuntimeError(f'Invalid sign string: {sign}')

def parse_effects(effect_list):
    if not isinstance(effect_list, list):
        raise RuntimeError(f'in parse_effects: expected list, got {effect_list}')

    result = []
    for effect in effect_list:
        match = re.match(
            r'^\s*(\w+)' # effect_type
            r'(?:\s*\((-?\d+)\))?' # effect_value
            r'(?:\s*(\w+)' # target_type
            r'\s*(?:\((-?\d+)\))?)?' # target_value
            r'([\w\s]*?)' # player_filter
            r'(?:\s*\|\s*([\w\s]+))?\s*$', # card_filter
            effect
        )
        if not match:
            raise RuntimeError(f'Invalid effect string: {effect}')

        effect_type = match.group(1)
        effect_value = match.group(2)
        target_type = match.group(3)
        target_value = match.group(4)
        player_filter = match.group(5)
        card_filter = match.group(6)

        if player_filter and target_type not in ('player', 'conditional_player', 'players', 'card', 'extra_card', 'cards'):
            raise RuntimeError(f'Invalid effect string: {effect}\nPlayer filter not allowed with {target_type}')
            
        if card_filter and target_type not in ('card', 'extra_card', 'cards'):
            raise RuntimeError(f'Invalid effect string: {effect}\nCard filter not allowed with {target_type}')

        result.append({
            'target':           CppEnum('target_type', target_type) if target_type else None,
            'player_filter':    CppEnum('target_player_filter', player_filter) if player_filter else None,
            'card_filter':      CppEnum('target_card_filter', card_filter) if card_filter else None,
            'effect_value':     int(effect_value) if effect_value else None,
            'target_value':     int(target_value) if target_value else None,
            'type':             CppEnum('effect_type', effect_type)
        })
    return result

def parse_effect_simple(effect_list, value_name, type_name):
    if not isinstance(effect_list, list):
        raise RuntimeError(f'in parse_effect_simple: expected list, got {effect_list}')

    result = []
    for effect in effect_list:
        match = re.match(
            r'^\s*(\w+)' # type
            r'(?:\s*\((-?\d+)\))?\s*$', # effect_value
            effect
        )
        if not match:
            raise RuntimeError(f'Invalid effect string: {effect}')
        
        effect_type = match.group(1)
        effect_value = match.group(2)

        result.append({
            value_name: int(effect_value) if effect_value else None,
            'type': CppEnum(type_name, effect_type)
        })
    return result

def parse_all_effects(card):
    try:
        return {
            'name':         card['name'] if 'name' in card else None,
            'image':        card['image'] if 'image' in card else None,
            'effects':      parse_effects(card['effects']) if 'effects' in card else None,
            'responses':    parse_effects(card['responses']) if 'responses' in card else None,
            'optionals':    parse_effects(card['optional']) if 'optional' in card else None,
            'equips':       parse_effect_simple(card['equip'], 'effect_value', 'equip_type') if 'equip' in card else None,
            'tags':         parse_effect_simple(card['tags'], 'tag_value', 'tag_type') if 'tags' in card else None,
            'expansion':    CppEnum('expansion_type', card['expansion']) if 'expansion' in card else None,
            'deck':         CppEnum('card_deck_type', card['deck']) if 'deck' in card else None,
            'modifier':     {'type': CppEnum('modifier_type', card['modifier'])} if 'modifier' in card else None,
            'mth_effect':   {'type': CppEnum('mth_type', card['mth_effect'])} if 'mth_effect' in card else None,
            'mth_response': {'type': CppEnum('mth_type', card['mth_response'])} if 'mth_response' in card else None,
            'equip_target': CppEnum('target_player_filter', card['equip_target']) if 'equip_target' in card else None,
            'color':        CppEnum('card_color_type', card['color']) if 'color' in card else None,
            'sign':         parse_sign(card['sign']) if 'sign' in card else None
        }
    except RuntimeError as error:
        raise RuntimeError(f"Error in card {card['name']}:\n{error}") from error

def add_expansion(card, name):
    if 'expansion' in card:
        card['expansion'] += ' ' + name
    else:
        card['expansion'] = name

def parse_file(data):
    result = {
        'deck': [],
        'characters': [],
        'goldrush': [],
        'highnoon': [],
        'fistfulofcards': [],
        'wildwestshow': [],
        'button_row': [],
        'stations': [],
        'train': [],
        'locomotive': [],
        'hidden': []
    }

    for card in data['main_deck']:
        card['deck'] = 'main_deck'
        if is_hidden(card):
            result['hidden'].append(parse_all_effects(card))
        else:
            for sign in card['signs']:
                card['sign'] = sign
                result['deck'].append(parse_all_effects(card))
    
    for card in data['character']:
        card['deck'] = 'character'
        result['characters'].append(parse_all_effects(card))
    
    for card in data['goldrush']:
        card['deck'] = 'goldrush'
        add_expansion(card, 'goldrush')
        count = card['count'] if 'count' in card else 1
        for _ in range(count):
            if is_hidden(card):
                result['hidden'].append(parse_all_effects(card))
            else:
                result['goldrush'].append(parse_all_effects(card))

    for name in ('highnoon', 'fistfulofcards', 'wildwestshow'):
        for card in data[name]:
            card['deck'] = name
            add_expansion(card, name)
            if is_hidden(card):
                result['hidden'].append(parse_all_effects(card))
            else:
                result[name].append(parse_all_effects(card))

    for card in data['station']:
        card['deck'] = 'station'
        add_expansion(card, 'greattrainrobbery')
        result['stations'].append(parse_all_effects(card))

    for card in data['train']:
        card['deck'] = 'train'
        add_expansion(card, 'greattrainrobbery')
        if is_hidden(card):
            result['hidden'].append(parse_all_effects(card))
        else:
            result['train'].append(parse_all_effects(card))
    
    for card in data['locomotive']:
        card['deck'] = 'locomotive'
        add_expansion(card, 'greattrainrobbery')
        result['locomotive'] .append(parse_all_effects(card))

    for card in data['button_row']:
        if is_hidden(card):
            result['hidden'].append(parse_all_effects(card))
        else:
            result['button_row'].append(parse_all_effects(card))

    return result

INCLUDE_FILENAMES = ['cards/card_data.h', 'cards/effect_enums.h']
OBJECT_DECLARATION = 'all_cards_t banggame::all_cards'

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print(f'Usage: {sys.argv[0]} bang_cards.yml bang_cards.cpp')
        sys.exit(1)

    with open(sys.argv[1], 'r', encoding='utf8') as file:
        bang_cards = parse_file(yaml.safe_load(file))
    
    with open(sys.argv[2], 'w', encoding='utf8') as file:
        print_cpp_file(bang_cards, OBJECT_DECLARATION,
            include_filenames=INCLUDE_FILENAMES,
            file=file)
